src/sleep/sleep.c

```diff
@@ -263,41 +263,91 @@ static int execute(
 }
 
 static int execute_s2h(const SleepConfig *sleep_config) {
-        _cleanup_close_ int tfd = -1;
-        struct itimerspec ts = {};
+        usec_t suspend_interval = sleep_config->hibernate_delay_sec, before_timestamp = 0, after_timestamp = 0;
+        bool woken_by_timer;
         int r;
 
         assert(sleep_config);
 
-        tfd = timerfd_create(CLOCK_BOOTTIME_ALARM, TFD_NONBLOCK|TFD_CLOEXEC);
-        if (tfd < 0)
-                return log_error_errno(errno, "Error creating timerfd: %m");
-
-        log_debug("Set timerfd wake alarm for %s",
-                  FORMAT_TIMESPAN(sleep_config->hibernate_delay_sec, USEC_PER_SEC));
+        while(!battery_is_low()) {
+                _cleanup_close_ int tfd = -1;
+                struct itimerspec ts = {};
+                int last_capacity = 0, current_capacity = 0, previous_discharge_rate, estimated_discharge_rate = 0;
+
+                tfd = timerfd_create(CLOCK_BOOTTIME_ALARM, TFD_NONBLOCK|TFD_CLOEXEC);
+                if (tfd < 0)
+                        return log_error_errno(errno, "Error creating timerfd: %m");
+
+                /* Store current battery capacity and current time before suspension */
+                r = read_battery_capacity_percentage();
+                if (r > 0) {
+                        last_capacity = r;
+                        log_debug("Current battery charge percentage: %d%%", last_capacity);
+                        before_timestamp = now(CLOCK_BOOTTIME);
+                } else if (r == -ENOENT)
+                        log_debug_errno(r, "Suspend Interval value set to %s: %m", FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
+                        /* In case of no battery, system will be suspended for hibernatedelaysec interval */
+                else
+                        return log_error_errno(r, "Error fetching battery capacity percentage: %m");
```

Comment:
Here r has battery_capacity in range of 0-100. So except -ENOENT, there is a rare possibility of any other error. In most cases the one of the first 2 condition would be true and since we need to check -ENOENT to have a verbose message so i think we should not remove this. Also we can not return in case of -ENOENT because then s2h will just exit and not do anything for devices with no battery.

Reply:
i have updated comment for better clarity

Reply:
And if we move the action performed when `r > 0` to the top-level, it would fail for device with no battery so that needs to be checked.
Should i change the sequence in which i am checking ?